### PR TITLE
feat!: update to tree-sitter 0.23

### PR DESCRIPTION
**Pull request by hendrikvanantwerpen**
Tree-sitter released version 0.23 recently. This contains a breaking, but ultimately very good change to how parser bindings work (https://github.com/tree-sitter/tree-sitter/pull/3069). The TLDR of that change is that parsers do not depend on tree-sitter anymore, but instead on a shard (and supposedly very stable) tree-sitter-language crate. As a result, clients are free to chose their tree-sitter version as the parser ABI is supported. Library and parser versions are less tighly coupled, and it should no longer be necessary to move all the parsers to the next tree-sitter version in lock-step to be able to upgrade the library.

Most of the changes in this PR are from running tree-sitter generate. I'll add comments on other changes I did to explain why I thought they are necessary, or where I'm not sure I did it correctly.

_Please review carefully, since this project has a non-standard setup, and I may have missed something._

## Complications with `MarkdownParser`

The Rust bindings, besides the gramamr, also provide a `MarkdownParser` implementation. It needs `tree-sitter` for its implementation. That means we cannot just remove the `tree-sitter` dependency like for most grammars.

Leaving the `tree-sitter` dependency as it was would negate the benefit of the switch to `tree-sitter-language` and still effectively pin the `tree-sitter` library for any user of this crate.

My proposal, as implemented in the PR, is to put the convenience parser behind a cargo feature 'parser'. Users that do not want it will get the benefits of the `tree-sitter-language` approach, while others will have to live (reasonably!) with the restriction on `tree-sitter` version they can use. 

The consequence is that users of the `MarkdownParser` will need to change their `Cargo.toml` to enable the `parser` feature on their `tree-sitter-md` dependency. Since they will probably need to change their `Cargo.toml` anyway when upgrading I don't think this is such a heavy burden, but it should be communicated clearly in release notes or something to avoid confusion.

Let me know if you think this is an acceptable solution!

### CODE REVIEW

File: bindings/rust/lib.rs [hunk 1]
```diff
@@ -55,354 +46,372 @@ pub const NODE_TYPES_BLOCK: &str = include_str!("../../tree-sitter-markdown/src/
...
-///
-/// This exposes the same methdos as [`TreeCursor`], but abstracts away the
-/// double block / inline structure of [`MarkdownTree`].
-pub struct MarkdownCursor<'a> {
-    markdown_tree: &'a MarkdownTree,
-    block_cursor: TreeCursor<'a>,
-    inline_cursor: Option<TreeCursor<'a>>,
-}
+#[cfg(feature = "parser")]
+mod parser {
```
Comment (ObserverOfTime) [inline bindings/rust/lib.rs]: Perhaps this should be moved to a separate file.

File: common/common.mak [hunk 2]
```diff
@@ -13,7 +13,7 @@ ifeq ($(shell echo $(PARSER_URL) | grep '^[a-z][-+.0-9a-z]*://'),)
 endif
 endif
 
-TS ?= tree-sitter
+TS ?= npx tree-sitter-cli
```
Comment (ObserverOfTime) [inline common/common.mak:16]: Nope.
```suggestion
TS ?= tree-sitter
```

File: package.json [hunk 3]
```diff
@@ -29,10 +29,10 @@
   },
   "devDependencies": {
     "prebuildify": "^6.0.0",
-    "tree-sitter-cli": "^0.22.2"
+    "tree-sitter-cli": "^0.23.0"
   },
   "peerDependencies": {
-    "tree-sitter": "^0.21.1"
+    "tree-sitter": "^0.21.0"
```
Comment (ObserverOfTime) [inline package.json]:
```suggestion
    "tree-sitter": "^0.21.1"
```

File: scripts/build.js [hunk 4]
```diff
@@ -5,7 +5,7 @@ const { join } = require("path");
 
 for (const dir of ["tree-sitter-markdown", "tree-sitter-markdown-inline"]) {
   console.log(`building ${dir}`);
-  execSync("tree-sitter generate --no-bindings", {
+  execSync("npx tree-sitter-cli generate --no-bindings", {
```
Comment (ObserverOfTime) [inline scripts/build.js]:
```suggestion
  execSync("tree-sitter generate --no-bindings", {
```

File: scripts/test.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ const parsers = ["tree-sitter-markdown", "tree-sitter-markdown-inline"];
 for (const dir of parsers) {
   console.log(`testing ${dir}`);
   try {
-    execSync("tree-sitter test", {
+    execSync("npx tree-sitter-cli test", {
```
Comment (ObserverOfTime) [inline scripts/test.js]:
```suggestion
    execSync("tree-sitter test", {
```

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]
Remove the `/src` files.

File: bindings/rust/lib.rs (hunk 1 above)
Comment (hendrikvanantwerpen) [reply bindings/rust/lib.rs]: Applied in 4542327d7e47133cb8065c6bb40ad4b13217a2b1.

File: common/common.mak (hunk 2 above)
Comment (hendrikvanantwerpen) [reply common/common.mak:16]: I have reverted these changes as you asked.

File: bindings/rust/lib.rs [hunk 6]
```diff
@@ -414,13 +60,21 @@ mod tests {
...
         inline_parser
-            .set_language(&inline_language())
+            .set_language(&INLINE_LANGUAGE.into())
             .expect("Error loading Markdown language");
     }
+}
+
+#[cfg(test)]
+#[cfg(feature = "parser")]
+mod parser_tests {
```
Comment (ObserverOfTime) [inline bindings/rust/lib.rs]: Move the parser tests to the other file too.

Comment (hendrikvanantwerpen) [reply bindings/rust/lib.rs]: 🤦🏻

Comment (hendrikvanantwerpen) [reply bindings/rust/lib.rs]: In f51080b376803e4adce9a5fb9c8b9f0ebbff4762.

**Review by ObserverOfTime — APPROVED** [review]

**Comment by ObserverOfTime** [comment]
Thank you!

**Comment by hendrikvanantwerpen** [comment]
@ObserverOfTime Thanks for merging these changes! Is it possible to do a versioned release of these as well?

**Pull request merged**